src/Razor/src/Microsoft.VisualStudio.LanguageServices.Razor/LanguageClient/Cohost/CohostDocumentPullDiagnosticsEndpoint.cs

```diff
@@ -124,13 +121,13 @@ public ImmutableArray<Registration> GetRegistrations(VSInternalClientCapabilitie
 
     private async Task<LspDiagnostic[]> GetCSharpDiagnosticsAsync(TextDocument razorDocument, CancellationToken cancellationToken)
     {
-        // TODO: This code will not work when the source generator is hooked up.
-        //       How do we get the source generated C# document without OOP? Can we reverse engineer a file path?
-        var projectKey = razorDocument.Project.ToProjectKey();
-        var csharpFilePath = _filePathService.GetRazorCSharpFilePath(projectKey, razorDocument.FilePath.AssumeNotNull());
-        // We put the project Id in the generated document path, so there can only be one document
-        if (razorDocument.Project.Solution.GetDocumentIdsWithFilePath(csharpFilePath) is not [{ } generatedDocumentId] ||
-            razorDocument.Project.GetDocument(generatedDocumentId) is not { } generatedDocument)
+        if (!razorDocument.TryComputeHintNameFromRazorDocument(out var hintName))
+        {
+            return [];
+        }
+
+        var generatedDocuments = await razorDocument.Project.GetSourceGeneratedDocumentsAsync(cancellationToken);
```

Comment:
There isn't currently a way around just iterating through all source generated documents sadly. There aren't many endpoints that have to run in devenv (just diagnostics and code actions I think?) so hopefully this is going to be an approach that is used sparingly

Reply:
Looks @chsienki had the same issues: https://github.com/dotnet/razor/blob/main/src/Razor/src/Microsoft.CodeAnalysis.Remote.Razor/ProjectSystem/RemoteProjectSnapshot.cs#L179-L186

Will unify these together into a helper somewhere.